// firebase-tools/lib/emulator/controller.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const _ = require("lodash");
const clc = require("cli-color");
const fs = require("fs");
const path = require("path");
const tcpport = require("tcp-port-used");
const utils = require("../utils");
const track = require("../track");
const registry_1 = require("../emulator/registry");
const types_1 = require("../emulator/types");
const constants_1 = require("../emulator/constants");
const functionsEmulator_1 = require("../emulator/functionsEmulator");
const databaseEmulator_1 = require("../emulator/databaseEmulator");
const firestoreEmulator_1 = require("../emulator/firestoreEmulator");
const hostingEmulator_1 = require("../emulator/hostingEmulator");
const error_1 = require("../error");
const getProjectId = require("../getProjectId");
const pubsubEmulator_1 = require("./pubsubEmulator");
exports.VALID_EMULATOR_STRINGS = types_1.ALL_EMULATORS;
function checkPortOpen(port, host) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const inUse = yield tcpport.check(port, host);
            return !inUse;
        }
        catch (e) {
            return false;
        }
    });
}
exports.checkPortOpen = checkPortOpen;
function waitForPortClosed(port, host) {
    return __awaiter(this, void 0, void 0, function* () {
        const interval = 250;
        const timeout = 30000;
        try {
            yield tcpport.waitUntilUsed(port, interval, timeout);
        }
        catch (e) {
            throw new error_1.FirebaseError(`TIMEOUT: Port ${port} on ${host} was not active within ${timeout}ms`);
        }
    });
}
exports.waitForPortClosed = waitForPortClosed;
function startEmulator(instance) {
    return __awaiter(this, void 0, void 0, function* () {
        const name = instance.getName();
        const info = instance.getInfo();
        track("emulators:start", name);
        const portOpen = yield checkPortOpen(info.port, info.host);
        if (!portOpen) {
            yield cleanShutdown();
            utils.logWarning(`Port ${info.port} is not open, could not start ${name} emulator.`);
            utils.logBullet(`To select a different port for the emulator, update your "firebase.json":
    {
      // ...
      "emulators": {
        "${name}": {
          "port": "${clc.yellow("PORT")}"
        }
      }
    }`);
            return utils.reject(`Could not start ${name} emulator, port taken.`, {});
        }
        yield registry_1.EmulatorRegistry.start(instance);
    });
}
exports.startEmulator = startEmulator;
function cleanShutdown() {
    return __awaiter(this, void 0, void 0, function* () {
        utils.logBullet("Shutting down emulators.");
        for (const name of registry_1.EmulatorRegistry.listRunning()) {
            utils.logBullet(`Stopping ${name} emulator`);
            yield registry_1.EmulatorRegistry.stop(name);
        }
        return true;
    });
}
exports.cleanShutdown = cleanShutdown;
function filterEmulatorTargets(options) {
    let targets = exports.VALID_EMULATOR_STRINGS.filter((e) => {
        return options.config.has(e) || options.config.has(`emulators.${e}`);
    });
    if (options.only) {
        targets = _.intersection(targets, options.only.split(","));
    }
    return targets;
}
exports.filterEmulatorTargets = filterEmulatorTargets;
function shouldStart(options, name) {
    const targets = filterEmulatorTargets(options);
    return targets.indexOf(name) >= 0;
}
exports.shouldStart = shouldStart;
function startAll(options) {
    return __awaiter(this, void 0, void 0, function* () {
        const targets = filterEmulatorTargets(options);
        options.targets = targets;
        const projectId = getProjectId(options, true);
        utils.logLabeledBullet("emulators", `Starting emulators: ${targets.join(", ")}`);
        if (options.only) {
            const requested = options.only.split(",");
            const ignored = _.difference(requested, targets);
            for (const name of ignored) {
                utils.logWarning(`Not starting the ${clc.bold(name)} emulator, make sure you have run ${clc.bold("firebase init")}.`);
            }
        }
        if (shouldStart(options, types_1.Emulators.FUNCTIONS)) {
            const functionsAddr = constants_1.Constants.getAddress(types_1.Emulators.FUNCTIONS, options);
            const projectId = getProjectId(options, false);
            const functionsDir = path.join(options.config.projectDir, options.config.get("functions.source"));
            let inspectFunctions;
            if (options.inspectFunctions === true) {
                options.inspectFunctions = "9229";
            }
            if (options.inspectFunctions) {
                inspectFunctions = Number(options.inspectFunctions);
                if (isNaN(inspectFunctions) || inspectFunctions < 1024 || inspectFunctions > 65535) {
                    throw new error_1.FirebaseError(`"${options.inspectFunctions}" is not a valid value for the --inspect-functions flag, please pass an integer between 1024 and 65535.`);
                }
                utils.logLabeledWarning("functions", `You are running the functions emulator in debug mode (port=${inspectFunctions}). This means that functions will execute in sequence rather than in parallel.`);
            }
            const functionsEmulator = new functionsEmulator_1.FunctionsEmulator({
                projectId,
                functionsDir,
                host: functionsAddr.host,
                port: functionsAddr.port,
                debugPort: inspectFunctions,
            });
            yield startEmulator(functionsEmulator);
        }
        if (shouldStart(options, types_1.Emulators.FIRESTORE)) {
            const firestoreAddr = constants_1.Constants.getAddress(types_1.Emulators.FIRESTORE, options);
            const args = {
                host: firestoreAddr.host,
                port: firestoreAddr.port,
                projectId,
                auto_download: true,
            };
            const rulesLocalPath = options.config.get("firestore.rules");
            if (rulesLocalPath) {
                const rules = path.join(options.projectRoot, rulesLocalPath);
                if (fs.existsSync(rules)) {
                    args.rules = rules;
                }
                else {
                    utils.logWarning(`Firestore rules file ${clc.bold(rules)} specified in firebase.json does not exist, starting Firestore emulator without rules.`);
                }
            }
            else {
                utils.logWarning(`No Firestore rules file specified in firebase.json, using default rules.`);
            }
            const firestoreEmulator = new firestoreEmulator_1.FirestoreEmulator(args);
            yield startEmulator(firestoreEmulator);
            utils.logLabeledBullet(types_1.Emulators.FIRESTORE, `For testing set ${clc.bold(`${firestoreEmulator_1.FirestoreEmulator.FIRESTORE_EMULATOR_ENV}=${firestoreAddr.host}:${firestoreAddr.port}`)}`);
        }
        if (shouldStart(options, types_1.Emulators.DATABASE)) {
            const databaseAddr = constants_1.Constants.getAddress(types_1.Emulators.DATABASE, options);
            const args = {
                host: databaseAddr.host,
                port: databaseAddr.port,
                projectId,
                auto_download: true,
            };
            if (shouldStart(options, types_1.Emulators.FUNCTIONS)) {
                const functionsAddr = constants_1.Constants.getAddress(types_1.Emulators.FUNCTIONS, options);
                args.functions_emulator_host = functionsAddr.host;
                args.functions_emulator_port = functionsAddr.port;
            }
            const rulesLocalPath = options.config.get("database.rules");
            if (rulesLocalPath) {
                const rules = path.join(options.projectRoot, rulesLocalPath);
                if (fs.existsSync(rules)) {
                    args.rules = rules;
                }
                else {
                    utils.logWarning(`Database rules file ${clc.bold(rules)} specified in firebase.json does not exist, starting Database emulator without rules.`);
                }
            }
            else {
                utils.logWarning(`No Database rules file specified in firebase.json, using default rules.`);
            }
            const databaseEmulator = new databaseEmulator_1.DatabaseEmulator(args);
            yield startEmulator(databaseEmulator);
            utils.logLabeledBullet(types_1.Emulators.DATABASE, `For testing set ${clc.bold(`${databaseEmulator_1.DatabaseEmulator.DATABASE_EMULATOR_ENV}=${databaseAddr.host}:${databaseAddr.port}`)}`);
        }
        if (shouldStart(options, types_1.Emulators.HOSTING)) {
            const hostingAddr = constants_1.Constants.getAddress(types_1.Emulators.HOSTING, options);
            const hostingEmulator = new hostingEmulator_1.HostingEmulator({
                host: hostingAddr.host,
                port: hostingAddr.port,
                options,
            });
            yield startEmulator(hostingEmulator);
        }
        if (shouldStart(options, types_1.Emulators.PUBSUB)) {
            if (!projectId) {
                throw new error_1.FirebaseError("Cannot start the Pub/Sub emulator without a project: run 'firebase init' or provide the --project flag");
            }
            const pubsubAddr = constants_1.Constants.getAddress(types_1.Emulators.PUBSUB, options);
            const pubsubEmulator = new pubsubEmulator_1.PubsubEmulator({
                host: pubsubAddr.host,
                port: pubsubAddr.port,
                projectId,
                auto_download: true,
            });
            yield startEmulator(pubsubEmulator);
        }
        const running = registry_1.EmulatorRegistry.listRunning();
        for (const name of running) {
            const instance = registry_1.EmulatorRegistry.get(name);
            if (instance) {
                yield instance.connect();
            }
        }
    });
}
exports.startAll = startAll;
